bundles/org.openhab.core.io.rest.sitemap/src/main/java/org/openhab/core/io/rest/sitemap/SitemapSubscriptionService.java

```diff
@@ -281,19 +317,41 @@ private EList<Widget> collectWidgets(String sitemapName, String pageId) {
         return widgets;
     }
 
+    private EList<Widget> collectAllWidgetsInSitemap(String sitemapName) {
+        EList<Widget> widgets = new BasicEList<>();
+        Sitemap sitemap = getSitemap(sitemapName);
+        if (sitemap != null) {
+            widgets = itemUIRegistry.getChildren(sitemap);
+            LinkedList<Widget> childrenQueue = new LinkedList<>(widgets);
+            while (!childrenQueue.isEmpty()) {
+                Widget child = childrenQueue.remove(0);
+                if (child instanceof LinkableWidget) {
+                    List<Widget> subWidgets = itemUIRegistry.getChildren((LinkableWidget) child);
+                    widgets.addAll(subWidgets);
+                    childrenQueue.addAll(subWidgets);
+                }
+            }
+        }
+        return widgets;
+    }
+
     private void removeCallbackFromListener(String sitemapPage, SitemapSubscriptionCallback callback) {
         ListenerRecord oldListener = pageChangeListeners.get(sitemapPage);
         if (oldListener != null) {
-            oldListener.pageChangeListener().removeCallback(callback);
-            if (!pageOfSubscription.containsValue(sitemapPage)) {
+            oldListener.widgetsChangeListener().removeCallback(callback);
+            if (!scopeOfSubscription.containsValue(sitemapPage)) {
                 // no other callbacks are left here, so we can safely dispose the listener
                 oldListener.serviceRegistration().unregister();
                 pageChangeListeners.remove(sitemapPage);
             }
         }
     }
 
-    private String getValue(String sitemapName, String pageId) {
+    private String getScopeIdentifier(String sitemapName) {
+        return sitemapName;
+    }
+
+    private String getScopeIdentifier(String sitemapName, String pageId) {
         return sitemapName + SITEMAP_PAGE_SEPARATOR + pageId;
     }
```

Comment:
```suggestion
    private String getScopeIdentifier(String sitemapName, @Nullable String pageId) {
        return pageId == null ? sitemapName : sitemapName + SITEMAP_PAGE_SEPARATOR + pageId;
    }
```

I think you can also collapse both versions of `addCallbackToListener` in one method, calling either `collectAllWidgetsInSitemap` or `collectWidgetsOnPage` (depending on `pageId == null`).